test/e2e/gateway/gateway_test.go

```diff
@@ -0,0 +1,192 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build e2e
+
+package gateway
+
+import (
+	"context"
+	"testing"
+
+	"github.com/projectcontour/contour/test/e2e"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gatewayv1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+type testGroup struct {
+	name    string
+	gateway *gatewayv1alpha1.Gateway
+	tests   map[string]func(t *testing.T, f *e2e.Framework)
+}
+
+// getTests defines the test groups that make up the Gateway test suite.
+// Each group shares a single Gateway definition.
+func getTests() []*testGroup {
+	var testGroups []*testGroup
+
+	testGroups = append(testGroups, &testGroup{
```

Comment:
shameless plug: this is where Ginkgo `BeforeEach` blocks in a `Describe` or `Context` would shine, so the gateway that needs to be used with a group of tests is close to the tests themselves and grouped together

Ginkgo does still end up with long test files though which is a pain, but that could be worked around with the structure we have here with helper methods that implement the tests themselves I think

Reply:
:+1: I'll spend some time looking at moving into Ginkgo so we can decide if we want to go that route. Also need to get Contour config file updating implemented, since ideally the Gateway tests would use a Gateway outside of the `projectcontour` namespace.

Reply:
Here's a branch that converts the Gateway tests to use Ginkgo (haven't bothered to convert subtests to use Gomega instead of testify):

https://github.com/skriss/contour/commit/a68d44134c37781b55791d0ca08459d7f1c7b6f7
gateway_test.go: https://github.com/skriss/contour/blob/a68d44134c37781b55791d0ca08459d7f1c7b6f7/test/e2e/gateway/gateway_test.go

Reply:
yeah i think IMO thats not bad and looks maybe easy to transition to without that much headache? i didnt do the work though so would have to ask you how you felt about it 

we could even pull the `It` blocks out into the individual test functions as well, and if needed probably even implement them as contexts themselves e.g.

```
	Describe("Insecure (Non-TLS) Gateway", func() {
		var gateway *gatewayv1alpha1.Gateway

		BeforeEach(func() {
...

		someTestName()
```

and 

```
func someTestName() {
    It("foo ...
```

or

```
func someTestName() {
    Context("foo ...
        BeforeEach(func() { ...
```

Reply:
yeah i think IMO thats not bad and looks maybe easy to transition to without that much headache? i didnt do the work though so would have to ask you how you felt about it 

we could even pull the `It` blocks out into the individual test functions as well, and if needed probably even implement them as contexts themselves e.g.

```
	Describe("Insecure (Non-TLS) Gateway", func() {
		var gateway *gatewayv1alpha1.Gateway

		BeforeEach(func() {
...

		someTestName()
```

and 

```
func someTestName() {
    It("foo ...
```

or

```
func someTestName() {
    Context("foo ...
        BeforeEach(func() { ...
```

Reply:
Yeah it's easy to switch to. I'm happy to move ahead with that approach if folks like using Ginkgo. I do need to figure out running the spec in parallel - haven't done that before - but it seems doable. The other question would be if we want to switch the subtests to use gomega instead of testify? We don't have to, we can mix and match, I think it's really just up to our preferences.

Reply:
we could even eventually do:

```
// some repeated test setup helper
func withTestNamespace(ns string, body func()) {
    Context("with namespace: " + ns, func() {
        BeforeEach(func() {
            // create ns
        ...
        AfterEach(func() {
            // delete ns
        ...
        // call inner test body
        body()
```

and

```
func someTestName() {
    withTestNamespace("some-test-namespace", func() {
        Context("foo ...
            BeforeEach(func() { ...
            It("does something", func() { ...
```

Reply:
if we invoke tests with the ginkgo CLI there should be a flag to control running in parallel and choosing the number of test nodes, i forget if it directly translates to a go test flag or not but shouldnt be too hard to work out

Reply:
WDYT about merging this as-is so we have the basic Gateway structure set up so any new/updated tests can be written there, then I'll get everything migrated over to Ginkgo in the next PR, then I'll continue with test case migrations?

Reply:
yeah that sounds good to me, don't want to derail your progress on this and other PRs

we might see some other patterns in PRs that follow that we want to have solutions for too 👍🏽